### PR TITLE
feat(nuxt): upgrade nitro + reduce node-specific usage

### PR DESCRIPTION
**Pull request by pi0**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR upgrades h3 and nitropack to the latest minor versions

Release notes (draft):
- https://github.com/unjs/h3/issues/486
- https://github.com/unjs/nitro/issues/1526

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: test/bundle.test.ts [hunk 1]
```diff
@@ -55,27 +57,42 @@ describe.skipIf(process.env.SKIP_BUNDLE_SIZE === 'true' || process.env.ECOSYSTEM
...
         "cookie-es",
         "debug",
         "defu",
         "destr",
         "devalue",
         "estree-walker",
+        "fill-range",
+        "glob-parent",
         "h3",
+        "h3/node_modules/iron-webcrypto",
```
Comment (pi0) [inline test/bundle.test.ts:73]: This seems wrong

File: test/bundle.test.ts [hunk 2]
```diff
@@ -92,45 +108,73 @@ describe.skipIf(process.env.SKIP_BUNDLE_SIZE === 'true' || process.env.ECOSYSTEM
...
+    expect.soft(roundToKilobytes(modules.totalBytes)).toMatchInlineSnapshot('"2564k"')
 
     const packages = modules.files
       .filter(m => m.endsWith('package.json'))
       .map(m => m.replace('/package.json', '').replace('node_modules/', ''))
       .sort()
     expect(packages).toMatchInlineSnapshot(`
       [
+        "@babel/parser",
+        "@nuxt/devalue",
```
[the remark this answers is not in the document]
Comment (pi0) [reply test/bundle.test.ts:123]: I think we might either increase score of valid npm packages to reduce behavior changes or use a function filter that calls first.

[the remark this answers is not in the document]
Comment (pi0) [reply test/bundle.test.ts:123]: (adding for reference) Current conflicts (both inline + external matchers and external wins which should not)

```
/nuxt/node_modules/.pnpm/@nuxt+ui-templates@1.3.1/node_modules/@nuxt/ui-templates/dist/templates/error-500.mjs 
{ inlineMatch: { [Function: match(@nuxt/ui-templates)] score: 18 },
  externalMatch: { [Function: match(/Users/pooya/Code/nuxt/node_modules)] score: 35 } } 

/nuxt/node_modules/.pnpm/@nuxt+devalue@2.0.2/node_modules/@nuxt/devalue/dist/devalue.mjs 
{ inlineMatch: { [Function: match(@nuxt/)] score: 6 },
  externalMatch: { [Function: match(/Users/pooya/Code/nuxt/node_modules)] score: 35 } } 

/nuxt/node_modules/.pnpm/@unhead+vue@1.2.2_vue@3.3.4/node_modules/@unhead/vue/dist/index.mjs
{ inlineMatch: { [Function: match(@unhead/vue)] score: 11 },
  externalMatch: { [Function: match(/Users/pooya/Code/nuxt/node_modules)] score: 35 } }  
```

[the remark this answers is not in the document]
Comment (pi0) [reply test/bundle.test.ts:123]: Okay that solved. the `chokidar` (and family!) extra packages coming from unstorage. We enabled new permanent `.data` dir for production by default. I am thinking to conditionally disable watching feature in unstorage for production somehow...

[the remark this answers is not in the document]
Comment (danielroe) [reply test/bundle.test.ts:123]: nice find.

**Review by danielroe — APPROVED** [review]

File: test/bundle.test.ts [hunk 3]
```diff
@@ -56,26 +56,22 @@ describe.skipIf(process.env.SKIP_BUNDLE_SIZE === 'true' || process.env.ECOSYSTEM
...
         "@vue/shared",
         "cookie-es",
-        "debug",
         "defu",
         "destr",
         "devalue",
         "estree-walker",
         "h3",
-        "has-flag",
+        "h3-nightly",
```
Comment (danielroe) [inline test/bundle.test.ts:64]: do we expect this? @pi0

Comment (pi0) [reply test/bundle.test.ts:64]: Yes. Newly supported in nitro it handles the symlink (they are not two packages). More info: https://github.com/unjs/nitro/pull/1553

**Comment by pi0** [comment]
locally confirmed image issue is h3 version by listhen dependency 😿 (nitro has to force version to it's expected one checking why is not)

File: docs/2.guide/2.directory-structure/1.server.md [hunk 4]
```diff
@@ -14,7 +14,7 @@ Nuxt automatically scans files inside these directories to register API and serv
 
 Each file should export a default function defined with `defineEventHandler()` or `eventHandler()` (alias).
 
-The handler can directly return JSON data, a `Promise` or use `event.node.res.end()` to send a response.
+The handler can directly return JSON data, a `Promise`, a `Response`, or use `event.node.res.end()` to send a response.
```
Comment (pi0) [inline docs/2.guide/2.directory-structure/1.server.md:17]:
```suggestion
The handler can directly return JSON data, a `Promise`, or use `event.node.res.end()` to send a response.
```

Comment (danielroe) [inline docs/2.guide/2.directory-structure/1.server.md]: might be worth instead mentioning `h3`'s `send` util?

Comment (pi0) [reply docs/2.guide/2.directory-structure/1.server.md]: send does not handles response (altohugh we might suggest `event.respondWith` with upcoming h3, i prefer we use direct responses. there is also typing concerns that we lose with response by advocating it...)

Comment (danielroe) [reply docs/2.guide/2.directory-structure/1.server.md]: Do we not support `Response` with new Nitro?

Comment (danielroe) [reply docs/2.guide/2.directory-structure/1.server.md]: I was suggesting dropping `event.node.res.end()`

Comment (pi0) [reply docs/2.guide/2.directory-structure/1.server.md]: Makes sense. (docs changes can be outside of this pr considering h3 is first to be released and available for all users before nuxt 3.7 we can update docs directly 👍🏼

File: test/bundle.test.ts [hunk 5]
```diff
@@ -55,33 +55,12 @@ describe.skipIf(process.env.SKIP_BUNDLE_SIZE === 'true' || process.env.ECOSYSTEM
...
-        "ms",
-        "node-fetch-native",
-        "ofetch",
-        "ohash",
-        "pathe",
-        "radix3",
-        "scule",
         "source-map-js",
-        "supports-color",
         "ufo",
```
Comment (danielroe) [inline test/bundle.test.ts:62]: interesting we have `hookable` and `ufo` here, though they are listed as runtime dependencies to-be-bundled [here](https://github.com/unjs/nitro/pull/1554/files).

Comment (pi0) [reply test/bundle.test.ts:62]: Yes, that’s something i have to investigate after upstream changes..

Comment (pi0) [reply test/bundle.test.ts:62]: BTW it makes sense to happen if we are externaling a lib that depends on them. (unhead maybe?)

Comment (pi0) [reply test/bundle.test.ts:62]: Post-inlinig is not possible.

I would recommend to inline known nuxt runtime dependencies as well to leverage same benefits of upstream. (or disable this optimization totally up to you!)

**Comment by pi0** [comment]
Only if we could make Node.js 18 mainline...

image

**Review by danielroe — APPROVED** [review]

**Pull request merged**